### PR TITLE
Always use cni unless running with dockershim

### LINKED ISSUE #14760: Default CNI to enabled, even for Docker runtime
With the release of cri-dockerd 0.2.5, it now defaults to CNI... So it should be recommended, for _all_ runtimes and single-node.

Previously it was only enabled for runtimes using CRI, such as containerd and cri-o, or multi-node. But now, also for docker.

----

This also applies to "none" driver, which has some weird special case (breaking containerd!)

From before, the default CNI plugin is "kindnet" for KIC and multi-node and "bridge" otherwise.

To be conservative with ancient versions, it can use the dockershim change point (1.24.0-alpha.2)

It was suggested before: #6094 (to use CNI), but then it was regarded as overkill (similar to CRI)

----

* ~~https://github.com/kubernetes/minikube/issues/2207~~
* ~~https://github.com/kubernetes/minikube/issues/3567~~

Basically, rules and interfaces only applied to "others"...

Now `crictl` and `/opt/cni/bin` are needed **everywhere**.

### PR DESCRIPTION
**Pull request by afbjorklund**
Since cri-dockerd 0.2.5, it now defaults to cni.

So when using cri (not dockershim), use cni too.

----

Closes #14760

### CODE REVIEW

**Comment by medyagh** [comment]
Thank you for the PR
@afbjorklund mind taking a look at this failure to see if it is related to this PR or not ?
https://storage.googleapis.com/minikube-builds/logs/14780/25250/Docker_Linux.html#fail_TestNetworkPlugins%2fgroup%2fauto%2fHairPin

File: pkg/minikube/cni/cni.go [hunk 1]
```diff
@@ -132,42 +132,44 @@ func IsDisabled(cc config.ClusterConfig) bool {
...
 
-func chooseDefault(cc config.ClusterConfig) Manager {
+func chooseDefault(cc config.ClusterConfig) (Manager, error) {
 	// For backwards compatibility with older profiles using --enable-default-cni
 	if cc.KubernetesConfig.EnableDefaultCNI {
 		klog.Infof("EnableDefaultCNI is true, recommending bridge")
-		return Bridge{}
-	}
-
-	if driver.BareMetal(cc.Driver) {
```
Comment (medyagh) [inline pkg/minikube/cni/cni.go:148]: is this change related to this PR ?

**Comment by afbjorklund** [comment]
> mind taking a look at this failure to see if it is related to this PR or not ?

it is related, it was hardcoding "not Docker" in the expectations.

`                        {"auto", []string{}, "", "", ContainerRuntime() != "docker"},`

File: pkg/minikube/cni/cni.go (hunk 1 above)
Comment (afbjorklund) [reply pkg/minikube/cni/cni.go:148]: yeah, it removed some weird special case for "none" driver - which needs CNI, just like all other drivers

**Comment by spowelljr** [comment]
@afbjorklund prow is failing to start

https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/minikube/14780/pull-minikube-build/1565039962942869504

**Comment by mgabeler-lee-6rs** [comment]
With the latest release of `crictl`, minikube with the `none` driver no longer works without a newer `cri-dockerd`, which doesn't work without this PR, and I believe #14703.

Any outlook on when those two (and perhaps #14827 which is related) might be ready to merge?

**Comment by afbjorklund** [comment]
There is a new cri-dockerd 0.3.0 out, with support for Kubernetes 1.26.0. Hopefully it can include the other fixes... ?

**Comment by mgabeler-lee-6rs** [comment]
Hmm, at one point I needed some combination of this & the other two PRs I listed to address #14724 -- that specifically lists #14703 more than this one for that issue, but I'd been testing with a local build merging all 3. Perhaps some further changes in cri-dockerd have made that not critical now, as it _seems_ to work out of the box again if I grab the latest releases of minikube & cri-dockerd & crictl.

**Comment by maxb** [comment]
How will the changes here impact users like me, who just want to run with the `none` driver and no network plugin at all, for a simple single-node cluster on a development laptop? I currently do this by adding the `--network-plugin=` option to the `cri-dockerd` command line.

**Comment by afbjorklund** [comment]
> How will the changes here impact users like me, who just want to run with the `none` driver and no network plugin at all, for a simple single-node cluster on a development laptop? I currently do this by adding the `--network-plugin=` option to the `cri-dockerd` command line.

The no-op plugin (--network-plugin=) and kubenet has been removed from upstream (retro-actively), so since Kubernetes 1.24 it is now _mandatory_ to use CRI and CNI... But you can still use a simple single-node network if you want to, by using `--cni=bridge` (it is rather similar).

**Comment by maxb** [comment]
Currently I have a functional single node cluster with no CNI at all, using Kubernetes 1.26.0, minikube 1.28.0, cri-dockerd 0.3.0, passing `--network-plugin=` to cri-dockerd.

**Comment by afbjorklund** [comment]
If you really want it, it would be possible to make a special case for docker.

**Comment by afbjorklund** [comment]
Previously the code used `""` to mean Docker container runtime, or `""` to mean No-op network plugin.

But now they mean "use the default" (whatever that happens to be), so need explicit strings for the old values.

**Comment by maxb** [comment]
My main interest is that the learning curve towards operating a single node driver=none setup, should avoid extra complexity where practical. It has already grown a fair bit recently, (cri-dockerd, crictl, needing to research the requirement to add `--network-plugin=`) but I think there's still value in allowing people to leave out CNI where not necessary, and defer tangling with that until a later stage of their Kubernetes educational journey.

I recognise driver=none is never going to be the preferred option for "just make it work, as quickly as possible" use-cases, but it really is great for people willing to put in a minor additional setup cost, in exchange for easier educational tinkering.

**Comment by afbjorklund** [comment]
> My main interest is that the learning curve towards operating a single node driver=none setup, should avoid extra complexity where practical. It has already grown a fair bit recently, (cri-dockerd, crictl, needing to research the requirement to add `--network-plugin=`) but I think there's still value in allowing people to leave out CNI where not necessary, and defer tangling with that until a later stage of their Kubernetes educational journey.

The simple days of Docker are _gone_, upstream has decided to make CRI mandatory (in 1.24) and to make CNI the default (in 1.25, retroactively also for 1.24). And on the latest release day, Docker was not even available (for 1.26).

So unfortunately people now **have** to learn both CRI and CNI, before the cluster will even start. It is possible that this will also include CSI in the future, but for now storage remains "optional" for a single-node cluster (and NFS still works...)

I fully assume that we will have to change to containerd as the default for 1.27, simple because dockerd isn't tested.
That will make life even worse, since you have to learn about containerd and buildkitd and other low-level stuff.

The `nerdctl` tools makes them slightly more comfortable to use, but it is not fully compatible with the old tools ?
You can also go the `podman` route (with CRI-O), if you prefer to use something "different". But still not Docker.

> I recognise driver=none is never going to be the preferred option for "just make it work, as quickly as possible" use-cases, but it really is great for people willing to put in a minor additional setup cost, in exchange for easier educational tinkering.

The "none" driver is used in both learning and testing environments, and it remains an important target.

I detailed some of the **worst** current bugs with it here:

* https://github.com/kubernetes/minikube/issues/15097

The lack of packaging and the lack of documentation makes it more cumbersome, than what it needed to be.

With Docker no longer being Open Source (with Desktop, and Machine no longer being developed), it is "gone".
I personally think that Lima is the best successor, after losing faith in Podman Desktop. Your mileage may vary.

So it seems that Kubernetes is now more Linux, in that you can tinker with it yourself or you can use a distribution ?
Hopefully, `minikube start` will continue to be such a Kubernetes distribution - with all the batteries included.

Anyway, _this_ feature (i.e. network-plugin) will remain - while available.

For the future of minikube, there is always the Slack channel: #minikube

**Review by medyagh — APPROVED** [review]

**Comment by medyagh** [comment]
@afbjorklund the only test failing is 


TestNetworkPlugins/group/false
--


<br class="Apple-interchange-newline">[TestNetworkPlugins/group/false](https://storage.googleapis.com/minikube-builds/logs/14780/28076/KVM_Linux.html#fail_TestNetworkPlugins%2fgroup%2ffalse)

could u take a look if we fix that we could merge

**Review by spowelljr — APPROVED** [review]

**Comment by tuanaac** [comment]
@afbjorklund could you please advise me how to change the subnet? 

> > How will the changes here impact users like me, who just want to run with the `none` driver and no network plugin at all, for a simple single-node cluster on a development laptop? I currently do this by adding the `--network-plugin=` option to the `cri-dockerd` command line.
> 
> The no-op plugin (--network-plugin=) and kubenet has been removed from upstream (retro-actively), so since Kubernetes 1.24 it is now _mandatory_ to use CRI and CNI... But you can still use a simple single-node network if you want to, by using `--cni=bridge` (it is rather similar).
> 
> ```yaml
@afbjorklund could you please advise me how to update the 
> {
>   "cniVersion": "0.3.1",
>   "name": "bridge",
>   "plugins": [
>     {
>       "type": "bridge",
>       "bridge": "bridge",
>       "addIf": "true",
>       "isDefaultGateway": true,
>       "forceAddress": false,
>       "ipMasq": true,
>       "hairpinMode": true,
>       "ipam": {
>           "type": "host-local",
>           "subnet": "10.244.0.0/16"
>       }
>     },
>     {
>       "type": "portmap",
>       "capabilities": {
>           "portMappings": true
>       }
>     }
>   ]
> }
> ```
